### PR TITLE
SOA-check: reject nxdomain response and check label of RR against qname

### PR DESCRIPTION
**Pull request by klaus3000**
Without these patches, PowerDNS even parses NXDOMAIN responses and falsely uses
wrong SOA records, i.e. the SOA record of a parent zone in authority section
when answering with NXDOMAIN.

Patch 1: ignore NXDOMAIN responses
Patch 2: only use SOA RR if the name of the RR matches the qname

### Short description
Fixes https://github.com/PowerDNS/pdns/issues/7058
It does not add all sanity checks, but it fixes issue 1 and 3. For issue 2 I do not found a solution as it seems all reponse sections (answer, authority, additional) are treated equally by the parser. Anyway, fixing issue 1+3 is sufficient.

### Checklist
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/resolver.cc [hunk 1]
```diff
@@ -271,17 +271,20 @@ bool Resolver::tryGetSOASerial(DNSName *domain, ComboAddress* remote, uint32_t *
...
     throw ResolverException("Query to '" + remote->toStringWithPort() + "' for SOA of '" + domain->toLogString() + "' returned wrong record type");
 
+  if(mdp.d_header.rcode != 0)
+    throw ResolverException("Query to '" + remote->toStringWithPort() + "' for SOA of '" + domain->toLogString() + "' returned Rcode " + RCode::to_s(mdp.d_header.rcode));
+
   *theirInception = *theirExpire = 0;
   bool gotSOA=false;
   for(const MOADNSParser::answers_t::value_type& drc :  mdp.d_answers) {
-    if(drc.first.d_type == QType::SOA) {
+    if(drc.first.d_type == QType::SOA && drc.first.d_name == *domain) {
```
[the remark this answers is not in the document]
Comment (klaus3000) [reply pdns/resolver.cc]: Indeed, this is not perfect, but the best "reference" to compare with I found in this function. At least it works better than before. Or where can I get the qname from the request?

I must admit that I totally not understand how the Inflighter works and where the response is mapped to the request: transaction id comparison, socket comparison, qname comparison (which is missing?)

[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/resolver.cc]: Ok, I looked into it a bit more, your fix is fine. `Resolver::tryGetSOASerial()` will parse the response and set the `domain`, `remote` and `id` to the received values, and then the `Inflighter` code will match these values against the ones it expects (via the `SlaveSenderReceiver::Identifier` which is a `std::tuple<DNSName, ComboAddress, uint16_t>`), and discard the response if there is no match.
So your fix is fine, you are correctly checking that the response was a `NoError` and that the `SOA` record matches the `qname`, and the matching to the query will be done later.

Comment (rgacogne) [inline pdns/resolver.cc]:
```suggestion
    if(drc.first.d_type == QType::SOA && drc.first.d_name == *domain && drc.first.d_place == DNSResourceRecord::ANSWER) {
```

File: pdns/resolver.cc [hunk 2]
```diff
@@ -271,17 +271,20 @@ bool Resolver::tryGetSOASerial(DNSName *domain, ComboAddress* remote, uint32_t *
...
       shared_ptr<SOARecordContent> src=getRR<SOARecordContent>(drc.first);
       if (src) {
         *theirSerial=src->d_st.serial;
         gotSOA = true;
       }
     }
-    if(drc.first.d_type == QType::RRSIG) {
+    if(drc.first.d_type == QType::RRSIG && drc.first.d_name == *domain) {
```
Comment (rgacogne) [inline pdns/resolver.cc]:
```suggestion
    if(drc.first.d_type == QType::RRSIG && drc.first.d_name == *domain && drc.first.d_place == DNSResourceRecord::ANSWER) {
```

**Pull request merged**